lib/fog/compute/google/requests/reset_windows_password.rb

```diff
@@ -0,0 +1,143 @@
+# Copyright 2015 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Changes:
+# March 2020: Modified example found here:
+# https://github.com/GoogleCloudPlatform/compute-image-windows/blob/master/examples/windows_auth_python_sample.py
+# to enable fog-google to change windows passwords.
+
+require "openssl"
+require "base64"
+require "json"
+
+module Fog
+  module Compute
+    class Google
+      class Mock
+        def reset_windows_password(_instance, _zone, _users)
+          Fog::Mock.not_implemented
+        end
+      end
+
+      class Real
+        ##
+        # Resets Windows passwords for users on Google's Windows based images.  Code based on Google provided example.
+        #
+        # @param instance [String] the name of the instance
+        # @param zone [String] the name of the zone of the instance
+        # @param user [String] the user whose password should be reset
+        #
+        # @return [String] new password
+        #
+        # @see https://cloud.google.com/compute/docs/instances/windows/automate-pw-generation
+        def reset_windows_password(instance, zone, user)
+         
+          # Pull the e-mail address of user authenticated to API 
+          email = @compute.request_options.authorization.issuer
+          
+          # Create a new key
+          key = OpenSSL::PKey::RSA.new(2048)
+          modulus, exponent = get_modulus_exponent_in_base64(key)
+
+					# Get Old Metadata
+          old_metadata = get_server(instance,zone).metadata
+          fingerprint = old_metadata.fingerprint
+
+          # Create JSON Object with needed information
+          metadata_entry = get_json_string(user, modulus, exponent, email)
+
+          # Create new metadata object
+          new_metadata = update_windows_keys(old_metadata, metadata_entry)
+
+          # Set metadata on instance
+          result = update_instance_metadata(instance, zone, fingerprint, new_metadata)
+
+          # Need to wait for new metadata to take effect.
```

Comment:
See the comment regarding using the `Server` model. If that doesn't work - please track the operation, something like:
```
          data = service.set_server_metadata(
            identity, zone_name, metadata[:fingerprint], new_metadata_items
          )
          operation = Fog::Compute::Google::Operations
                      .new(:service => service)
                      .get(data.name, data.zone)
          operation.wait_for { ready? } unless async
```
But I am in favour of not replicating this logic.

Reply:
There are two periods here we need to be concerned about.  First, the delay between setting the metadata and it taking affect, and second the delay between the metadata taking affect and the password being returned on serial output.

I was able to use the server model and can use the async=false.  When a machine has been booted for awhile, the second delay seems almost negligible.  However, when combined with booting a machine using `vagrant-google`, I'm needing to wait 30-60 seconds.  

I wonder if since this is using some Windows functions to create a user and reset a password, perhaps Windows is not fully ready even though the machine is ready?

My current thought is to have some retry logic where if serial port output is empty, wait and check again.

Reply:
Sorry, I think I missed this bit.

Did you find a way to check the serial port here or you think waiting 60 seconds is good enough for now and we should add an FR to track this?

Reply:
I had changed this to be a loop that rechecked every 5 seconds.  After looking at it again i realized it was an infinite loop.  If we never get output here, then the reset operation has failed.  Therefore, I added a loop counter and now raise an exception if we time out after 60 seconds.

Reply:
👌 Fantastic, thank you!